File: queue/ts/queue.ts
// --- Implementations ---

class Queue<T> {
  private elements: T[] = [];

  // Enqueue an element to the queue
  enqueue(element: T): void {
    this.elements.push(element);
  }

  // Add elements to the queue
  addAll(elements: T[]): void {
    this.elements.push(...elements);
  }

  // Dequeue an element from the queue
  dequeue(): T {
    if (this.isEmpty) throw new Error("Queue is empty");
    return this.elements.shift()!;
  }

  // Get the front element of the queue without removing it
  peek(): T {
    if (this.isEmpty) throw new Error("Queue is empty");
    return this.elements[0];
  }

  // Check if the queue is empty
  get isEmpty(): boolean {
    return this.elements.length === 0;
  }

  // Get the size of the queue
  get size(): number {
    return this.elements.length;
  }

  // Print the elements of the queue
  toString(): string {
    return `[${this.elements.join(", ")}]`;
  }
}

class QueueCircular<T> {
  private capacity: number;
  private defaultValue: T;
  private elements: T[];
  private front: number = 0;
  private rear: number = 0;
  private _size: number = 0;

  constructor(capacity: number, defaultValue: T) {
    this.capacity = capacity;
    this.defaultValue = defaultValue;
    this.elements = new Array(capacity).fill(defaultValue);
  }

  // Check if the queue is empty
  get isEmpty(): boolean {
    return this._size === 0;
  }

  // Check if the queue is full
  get isFull(): boolean {
    return this._size === this.capacity;
  }

  // Get the length of the queue
  get length(): number {
    return this._size;
  }

  // Enqueue an element to the queue
  enqueue(element: T): void {
    if (this.isFull) throw new Error("Queue is full");

    this.elements[this.rear] = element;
    this.rear = (this.rear + 1) % this.elements.length;
    this._size++;
  }

  // Dequeue an element from the queue
  dequeue(): T {
    if (this.isEmpty) throw new Error("Queue is empty");

    const element = this.elements[this.front];
    this.elements[this.front] = this.defaultValue;
    this.front = (this.front + 1) % this.elements.length;
    this._size--;

    return element;
  }

  // Get the front element of the queue without removing it
  peek(): T {
    if (this.isEmpty) throw new Error("Queue is empty");

    return this.elements[this.front];
  }

  // Print the elements of the queue
  toString(): string {
    return `[${this.elements.join(", ")}]`;
  }
}

// Example Usage
console.log("--- Linear Queue ---");
const queue = new Queue<number>();

// Enqueue elements
queue.enqueue(1);
queue.enqueue(2);
queue.enqueue(3);

// Print queue size
console.log(`Queue size: ${queue.size}`);

// Peek the front element
console.log(`Front element: ${queue.peek()}`);

// Dequeue elements and print them
while (!queue.isEmpty) {
  console.log(`Dequeued element: ${queue.dequeue()}`);
}

console.log("\n--- Circular Queue ---");
// Create a circular queue with a capacity of 4
const circularQueue = new QueueCircular<number>(4, -1);
circularQueue.enqueue(1);
circularQueue.enqueue(2);
console.log(circularQueue.toString());
circularQueue.enqueue(3);
circularQueue.enqueue(4);

// Print the front element
console.log(`Front element: ${circularQueue.peek()}`);

// Dequeue elements
console.log(`Dequeued element: ${circularQueue.dequeue()}`);
console.log(`Dequeued element: ${circularQueue.dequeue()}`);
console.log(circularQueue.toString());
circularQueue.enqueue(5);
console.log(circularQueue.toString());

File: queue/dart/queue.dart
void main() {
  Queue queue = Queue();

  // Enqueue elements
  queue.enqueue(1);
  queue.enqueue(2);
  queue.enqueue(3);

  // Print queue size
  print('Queue size: ${queue.size}');

  // Peek the front element
  print('Front element: ${queue.peek()}');

  // Dequeue elements and print them
  while (!queue.isEmpty) {
    print('Dequeued element: ${queue.dequeue()}');
  }

  // Create a circular queue with a capacity of 3
  QueueCircular circularQueue = QueueCircular(capacity: 4, defaultValue: -1);
  circularQueue.enqueue(1);
  circularQueue.enqueue(2);
  print(circularQueue);
  circularQueue.enqueue(3);
  circularQueue.enqueue(4);

  // Print the front element
  print('Front element: ${circularQueue.peek()}');

  // Dequeue elements
  print('Dequeued element: ${circularQueue.dequeue()}');
  print('Dequeued element: ${circularQueue.dequeue()}');
  print(circularQueue);
  circularQueue.enqueue(5);
  print(circularQueue);
}

class Queue<T> {
  List<T> _elements = [];

  // Enqueue an element to the queue
  void enqueue(T element) {
    _elements.add(element);
  }

  // Add elements to the queue
  addAll(Iterable<T> elements) {
    _elements.addAll(elements);
  }

  // Dequeue an element from the queue
  T dequeue() {
    if (isEmpty) throw StateError('Queue is empty');
    return _elements.removeAt(0);
  }

  // Get the front element of the queue without removing it
  T peek() {
    if (isEmpty) throw StateError('Queue is empty');

    return _elements.first;
  }

  // Check if the queue is empty
  bool get isEmpty => _elements.isEmpty;

  // Get the size of the queue
  int get size => _elements.length;

  // Print the elements of the queue
  @override
  String toString() {
    return _elements.toString();
  }
}

class QueueCircular<T> {
  final int _capacity;
  final T _defaultValue;
  List<T> _elements = [];
  int _front = 0;
  int _rear = 0;
  int _size = 0;

  QueueCircular({required int capacity, required T defaultValue})
      : _capacity = capacity,
        _defaultValue = defaultValue,
        _elements = List.filled(capacity, defaultValue);

  // Check if the queue is empty
  bool get isEmpty => _size == 0;

  // Check if the queue is full
  bool get isFull => _size == _capacity;

  // Get the length of the queue
  int get length => _size;

  // Enqueue an element to the queue
  void enqueue(T element) {
    if (isFull) throw StateError('Queue is full');

    _elements[_rear] = element;
    _rear = (_rear + 1) % _elements.length;
    _size++;
  }

  // Dequeue an element from the queue
  T dequeue() {
    if (isEmpty) throw StateError('Queue is empty');

    T element = _elements[_front];
    _elements[_front] = _defaultValue;
    _front = (_front + 1) % _elements.length;
    _size--;

    return element;
  }

  // Get the front element of the queue without removing it
  T peek() {
    if (isEmpty) throw StateError('Queue is empty');

    return _elements[_front];
  }

  // Print the elements of the queue
  @override
  String toString() {
    return _elements.toString();
  }
}
